proposal/new-js-importer.d.ts

```diff
@@ -0,0 +1,322 @@
+ * The canonical URL must be absolute, including a scheme, but the specific
+ * structure is up to the importer. In most cases, the stylesheet in question
+ * will exist on disk and the importer will just return a `file://` URL for it.
+ *
+ * The `canonicalize()` method takes a URL string that may be either relative or
+ * absolute. If the importer recognizes that URL, it returns a corresponding
+ * absolute URL (including a scheme). This is the _canonical URL_ for the
+ * stylesheet in question. Although the input URL may omit a file extension or
+ * an initial underscore, the canonical URL must be fully resolved.
+ *
+ * For a stylesheet that's loaded from the filesystem, the canonical URL will be
+ * the absolute `file:` URL of the physical file on disk. If it's generated
+ * in-memory, the importer should choose a custom URL scheme to guarantee that
+ * its canonical URLs don't conflict with any other importer's.
+ *
+ * Having a canonical URL for each stylesheet allows Sass to ensure that the
+ * same stylesheet isn't loaded multiple times in the new module system.
+ *
+ * ### Loading
+ *
+ * The second step actually loads the text of the stylesheet. The `load()`
+ * method takes a canonical URL that was returned by `canonicalize()` and
+ * returns the contents of the stylesheet at that URL. This is only called once
+ * per compilation for each canonical URL; future loads of the same URL will
+ * re-use either the existing module (for `@use` and `@forward`) or the parse
+ * tree (for `@import`).
+ *
+ * ### `FileImporter`
+ *
+ * This proposal also adds a special type of importer known as a `FileImporter`.
+ * This importer makes the common case of redirecting loads to somewhere on the
+ * physical filesystem easier. It doesn't require the caller to implement
+ * `load()`, since that's always going to be the same for files on disk.
+ *
+ * A `FileImporter` only needs to have one method: `findFileUrl()`, which takes
+ * a relative URL and returns the absolute `file:` URL it should refer to.
+ * However, this URL doesn't need to be fully canonicalized: the Sass compiler
+ * will take care of resolving partials, file extensions, index files, and so
+ * on.
+ */
+
+/** ## API */
+
+import {URL} from 'url';
+
+import {Options, Syntax} from './new-js-api';
+import './new-js-api';
+
+declare module './new-js-api' {
+  interface Options<sync extends 'sync' | 'async'> {
+    importers?: _Importer<sync>[];
+  }
+}
+
+/** This definition supersedes the one in the New JavaScript API proposal. */
+type StringOptions<sync extends 'sync' | 'async'> = Options<sync> & {
+  syntax?: Syntax;
+} & (
+    | {url?: URL}
+    | {
+        importer: _Importer<sync>;
+        url: URL;
+      }
+  );
+
+/**
+ * This interface represents an [importer]. When the importer is invoked with a
+ * string `string`:
+ *
+ * [importer]: ../spec/modules.md#importer
+ *
+ * * Let `fromImport` be `true` if the importer is being run for an `@import`
```

Comment:
Super nitty but IMO hyphens make lists more readable in jsdocs

```suggestion
 * - Let `fromImport` be `true` if the importer is being run for an `@import`
```